File: source/downloads/code/LeetCode/Java/Single-Number3.java
public class Solution {
        public int[] singleNumber(int[] nums) {
            int diff = 0;
            for(int num: nums){
                diff ^= num;
            }
            // Get last set bit
            diff &= -diff;
            
            int[] ret = new int[2];
            for(int num: nums){
                if((num & diff) == 0)
                    ret[0] ^= num;
                else
                    ret[1] ^= num;
            }
            
            return ret;
        }
}

File: downloads/code/LeetCode/Python/Single-Number3.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        # calculate the xor for all nums
        xor = 0
        result = [0,0]
        for i in nums:
            xor ^= i
        # get the low bit
        lowbit = xor & -xor

        # get the result
        for i in nums:
            if lowbit & i == 0:
                result[0] ^= i
            else:
                result[1] ^= i
        return result

        
